torch/distributions/truncated_distribution.py

```diff
@@ -0,0 +1,81 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all, _finfo
+
+
+class TruncatedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies truncation to base distribution.
+    """
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), sample_method='inversion', *args, **kwargs):
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
+        self.lower_bound, self.upper_bound = broadcast_all(lower_bound, upper_bound)
+        self.base_dist = base_distribution
+        cdf_low, cdf_high = self.base_dist.cdf(self.lower_bound), self.base_dist.cdf(self.upper_bound)
+        if sample_method in ['rejection', 'inversion']:
+            self.sample = {'rejection': self._rejection_sample, 'inversion': self._inversion_sample}[sample_method]
+        # we're going to use this value a lot
+        self.c = (cdf_high - cdf_low).log()
+
+    @constraints.dependent_property
+    def params(self):
+        return self.base_dist.params  # TODO add params of truncation
+
+    @constraints.dependent_property
+    def support(self):
+        # Note: The proper way to implement this is intersection([lower_bound, upper_bound], base_dist.support)
+        # This requires intersection method to be implemented for constraints.
+        return constraints.interval(self.lower_bound, self.upper_bound)
+
+    @property
+    def batch_shape(self):
+        return self.base_dist.batch_shape
+
+    @property
+    def event_shape(self):
+        return self.base_dist.event_shape
+
+    def _inversion_sample(self, sample_shape=torch.Size()):
```

Comment:
I'm inclined to implement `Inversion` and `Rejection` as different classes because the interfaces differ: `Inversion` allows reparametrization hence allows an `.rsample()` whereas `Rejection` is not reparametrizable and hence only implements `.sample()` (It can be  partially reparametrized via RSVI but that requires yet a different interface). Also, Pyro defines a different `Rejector` class to do rejection sampling given a more general rejection criterion.

Reply:
I'm inclined to omit rejection sampling from pytorch actually. It's hard to make it work efficiently oob for a range of distributions. What do you think?

Reply:
I agree, let's omit rejection sampling.